### PR TITLE
Endpoint

### PR DESCRIPTION
**Pull request by kopiczko**
Towards https://github.com/giantswarm/giantswarm/issues/1483

Adds Endpoint and WatchEndpoint methods to TPR object

### CODE REVIEW

File: tpr/kind.go [hunk 1]
```diff
@@ -38,3 +38,39 @@ func extractKindAndGroup(name string) (kind, group string, err error) {
 
 	return
 }
+
+// unpluralizedSuffixes is a list of resource suffixes that are the same plural
+// and singular This is only is only necessary because some bits of code are
```
Comment (asymmetric) [inline tpr/kind.go]: Repetition typo

File: tpr/tpr.go [hunk 2]
```diff
@@ -55,7 +54,9 @@ type TPR struct {
 	apiVersion  string // apiVersion is group/version
 	description string
 
-	endpoint string
+	// resourceName is name of a resource registered in Kubernetes API for
```
Comment (asymmetric) [inline tpr/tpr.go]: is the name

File: tpr/tpr.go [hunk 3]
```diff
@@ -107,6 +108,28 @@ func (t *TPR) Name() string { return t.name }
 // Group returns a TPR group extracted from Name. See Config.Name godoc for details.
 func (t *TPR) Group() string { return t.group }
 
+// Endpoint returns a TPR resurce endpoint registered in Kubernetes API under
```
Comment (asymmetric) [inline tpr/tpr.go]: in the K8s API

File: tpr/tpr.go [hunk 4]
```diff
@@ -107,6 +108,28 @@ func (t *TPR) Name() string { return t.name }
...
+// a given namespace. Default namespace will be used when argument is an empty
```
Comment (asymmetric) [inline tpr/tpr.go]: The default namespace ... when the argument ;)

File: tpr/tpr.go [hunk 5]
```diff
@@ -107,6 +108,28 @@ func (t *TPR) Name() string { return t.name }
...
+// string.
+func (t *TPR) Endpoint(namespace string) string {
+	nsResource := t.resourceName
+	if len(namespace) != 0 {
+		nsResource = "namespace/" + namespace + "/" + t.resourceName
+	}
+	return "/apis/" + t.group + "/" + t.version + "/" + nsResource
```
Comment (asymmetric) [inline tpr/tpr.go]: Can we use `strings.Join` here instead?

Comment (kopiczko) [inline tpr/tpr.go]: I thought about path.Join already. Problem is that we'd have to create different slice depending if namespace is `""` or not.

Comment (kopiczko) [inline tpr/tpr.go]: There are also tests to make sure there aren't missing or redundant slashes.

File: tpr/tpr.go [hunk 6]
```diff
@@ -119,9 +119,9 @@ func (t *TPR) Endpoint(namespace string) string {
 	return "/apis/" + t.group + "/" + t.version + "/" + nsResource
 }
 
-// Endpoint returns a TPR watch resurce endpoint registered in Kubernetes API
-// under a given namespace. Default namespace will be used when argument is an
-// empty string.
+// Endpoint returns a TPR watch resurce endpoint registered in the Kubernetes
```
Comment (asymmetric) [inline tpr/tpr.go]: resource

File: tpr/tpr.go [hunk 7]
```diff
@@ -119,9 +119,9 @@ func (t *TPR) Endpoint(namespace string) string {
...
+// API under a given namespace. Default namespace will be used when the
```
Comment (asymmetric) [inline tpr/tpr.go]: the default (above too)

File: tpr/kind.go [hunk 8]
```diff
@@ -40,8 +40,8 @@ func extractKindAndGroup(name string) (kind, group string, err error) {
 }
 
 // unpluralizedSuffixes is a list of resource suffixes that are the same plural
-// and singular This is only is only necessary because some bits of code are
-// lazy and don't actually use the RESTMapper like they should.
+// and singular. This is only necessary because some bits of [kubernetes] code
```
Comment (asymmetric) [inline tpr/kind.go]: Are the `[]` there for a link? Otherwise I'd convert to `()`.

**Comment by kopiczko** [comment]
@asymmetric 

> Might be worth extracting this pluralization stuff to its own library at some point.

Maybe. It's shitty at the moment :P And it must be like that to match K8s API. When TPR will be updated to [allow specifying custom plural form](https://github.com/giantswarm/giantswarm/issues/1483#issuecomment-303140694), that may make more sense as people maybe will be adding exceptions.

File: tpr/kind.go [hunk 9]
```diff
@@ -38,3 +38,39 @@ func extractKindAndGroup(name string) (kind, group string, err error) {
 
 	return
 }
+
+// unpluralizedSuffixes is a list of resource suffixes that are the same plural
```
Comment (rossf7) [inline tpr/kind.go]: Nitpick but I think "when plural and singular" is easier to read.

File: tpr/kind_test.go [hunk 10]
```diff
@@ -41,3 +41,32 @@ func TestExtractKindAndGroup(t *testing.T) {
 		assert.Equal(t, tc.expectedGroup, group, "#%d", i)
 	}
 }
+
+func TestUnsafeGuessKindToResource(t *testing.T) {
```
Comment (rossf7) [inline tpr/kind_test.go]: <3 that the tests cover all the exceptions.

File: tpr/tpr.go [hunk 11]
```diff
@@ -55,7 +54,9 @@ type TPR struct {
 	apiVersion  string // apiVersion is group/version
 	description string
 
-	endpoint string
+	// resourceName is the name of a resource registered in Kubernetes API
```
Comment (rossf7) [inline tpr/tpr.go]: "in the Kubernetes API"

File: tpr/tpr.go [hunk 12]
```diff
@@ -107,6 +108,28 @@ func (t *TPR) Name() string { return t.name }
...
+func (t *TPR) Endpoint(namespace string) string {
+	nsResource := t.resourceName
+	if len(namespace) != 0 {
+		nsResource = "namespace/" + namespace + "/" + t.resourceName
+	}
+	return "/apis/" + t.group + "/" + t.version + "/" + nsResource
+}
+
+// Endpoint returns a TPR watch resource endpoint registered in the Kubernetes
+// API under a given namespace. Default namespace will be used when the
```
Comment (rossf7) [inline tpr/tpr.go]: "The default namespace"

**Comment by kopiczko** [comment]
@rossf7 yes, but this is a library, it should also support 1.6 for a while.

**Comment by kopiczko** [comment]
Agh..., sorry for the commit message. Clicked before read..

**Pull request merged**